### PR TITLE
Support IP Type of Service / Traffic Class: --ip-tos

### PR DESCRIPTION
**Pull request by orgads**
Add `--ip-tos` option to the command line tool for setting TOS for IPv4 or Traffic Class for IPv6.

### CODE REVIEW

File: CMake/CurlTests.c
Comment (vszakats) [inline CMake/CurlTests.c:321]: It'd be useful to pre-cache this for Windows, which always seems to fail this test:
```diff
--- a/CMake/Platforms/WindowsCache.cmake
+++ b/CMake/Platforms/WindowsCache.cmake
@@ -170,6 +170,7 @@ set(HAVE_IOCTLSOCKET_CAMEL_FIONBIO 0)
 set(HAVE_IOCTLSOCKET_FIONBIO 1)
 set(HAVE_IOCTL_FIONBIO 0)
 set(HAVE_SETSOCKOPT_SO_NONBLOCK 0)
+set(HAVE_SETSOCKOPT_SOL_IP 0)
 set(HAVE_POSIX_STRERROR_R 0)
 set(HAVE_BUILTIN_AVAILABLE 0)
 set(HAVE_MSG_NOSIGNAL 0)
```

Comment (orgads) [reply CMake/CurlTests.c:321]: Nice, thanks!

**Comment by orgads** [comment]
I don't know what to do with the remaining tests that fail. Can you please advise?

**Comment by bagder** [comment]
Why is a new libcurl option needed? Can't `CURLOPT_SOCKOPTFUNCTION` already do this?

**Comment by bagder** [comment]
This is a IPv4-only feature. What happens for IPv6 sockets and how is a user supposed to know when this works or not?

Can you please explain your use case for this?

File: src/tool_getparam.c
Comment (bagder) [inline src/tool_getparam.c:1007]: The command line option parsing cannot depend on this. If it should, it needs to check if *this* libcurl supports it in run-time.

**Comment by orgads** [comment]
> Why is a new libcurl option needed? Can't `CURLOPT_SOCKOPTFUNCTION` already do this?

We use it with the command-line interface. And if I understand correctly, `CURLOPT_SOCKOPTFUNCTION` is for user applications, not for internal use by curl, right?

File: src/tool_getparam.c
Comment (orgads) [reply src/tool_getparam.c:1007]: The problem is that all the "user-friendly" values here are defined in `netinet/ip.h`, which is not available on all platforms. What do you suggest to do? Maybe support only numeric values on other platforms? But what for actually, if it has no effect?

**Comment by orgads** [comment]
> This is a IPv4-only feature. What happens for IPv6 sockets and how is a user supposed to know when this works or not?
> 
> Can you please explain your use case for this?

We need it for applying QoS for IP packets. All the other parts of our application support this, and we've been patching curl for long time to add support. Now we'd like to upstream our patches (2 more will follow :)).

I noticed that another used proposed a similar change in #7762, but it was abandoned. I took the user-friendly names from that PR.

**Comment by bagder** [comment]
> > Why is a new libcurl option needed? Can't `CURLOPT_SOCKOPTFUNCTION` already do this?
> 
> We use it with the command-line interface. And if I understand correctly, `CURLOPT_SOCKOPTFUNCTION` is for user applications, not for internal use by curl, right?

You add `CURLOPT_TYPE_OF_SERVICE` in this PR. It is not for the command line tool, that's a new option for libcurl and I wonder why you need it.

File: src/tool_getparam.c
Comment (bagder) [reply src/tool_getparam.c:1007]: We work hard to decouple the tool from the library and its specific versions and features. What if you run with a different build of the library - with the same tool version, how would the tool know what features the library supports?

**Comment by orgads** [comment]
> > > Why is a new libcurl option needed? Can't `CURLOPT_SOCKOPTFUNCTION` already do this?
> > 
> > 
> > We use it with the command-line interface. And if I understand correctly, `CURLOPT_SOCKOPTFUNCTION` is for user applications, not for internal use by curl, right?
> 
> You add `CURLOPT_TYPE_OF_SERVICE` in this PR. It is not for the command line tool, that's a new option for libcurl and I wonder why you need it.

It is used in `tool_operate.c`. If I understand the relations correctly, this sets the library options according to the command-line arguments. So do you suggest using `CURLOPT_SOCKOPTFUNCTION` there? I don't see any other option that uses it there, why should this one be different?

File: src/tool_getparam.c
Comment (orgads) [reply src/tool_getparam.c:1007]: If you prefer that, I'll just remove the readable names, and let it just accept a numeric value (that's what we use anyway. The idea to use names was taken from #7762). Then I can remove the condition.

Comment (orgads) [reply src/tool_getparam.c:1007]: By the way, `ECH` is also disabled in command-line with `#ifdef`. Is that wrong too?

File: src/tool_getparam.c
Comment (orgads) [reply src/tool_getparam.c]: Done

File: src/tool_getparam.c
Comment (orgads) [reply src/tool_getparam.c]: Done

**Comment by orgads** [comment]
> > So do you suggest using CURLOPT_SOCKOPTFUNCTION there?
> 
> I do. I don't see why libcurl needs another option when you can already accomplish the same thing using an existing option.

The next feature I plan to add is setting VLAN priority, which is another setsockopt (`setsockopt(sockfd, SOL_SOCKET, SO_PRIORITY, (char*)&priority, sizeof(priority))`). How can I support both options? There is only one `SOCKOPTFUNCTION` entry.

**Comment by bagder** [comment]
> There is only one SOCKOPTFUNCTION entry.

So do both in the same callback?

File: m4/curl-functions.m4
Comment (orgads) [reply m4/curl-functions.m4:59]: Thanks, done.

File: src/tool_getparam.c
Comment (orgads) [reply src/tool_getparam.c]: Done, I hope you like it now :)

**Comment by orgads** [comment]
> > There is only one SOCKOPTFUNCTION entry.
> 
> So do both in the same callback?

Done. Is this what you meant?

File: m4/curl-functions.m4
Comment (bagder) [inline m4/curl-functions.m4:3415]: Why check for `SOL_IP` support for setsockopt when the code is using `IPPROTO_IP` in that argument?

Comment (bagder) [reply m4/curl-functions.m4:3415]: According to some [posts](https://lists.gnu.org/archive/html/autoconf/2010-03/msg00119.html) I found, maybe the other way around is more portable?

Comment (orgads) [reply m4/curl-functions.m4]: Done.

File: src/tool_getparam.c
Comment (orgads) [inline src/tool_getparam.c:1034]: This is a bit ugly. Do you prefer to add a configure check (for `IP_TOS`? something else?), check here for `IP_TOS` or just remove these abbreviations altogether?

**Comment by orgads** [comment]
Now we also have Windows 10 support :)

According to [this answer](https://stackoverflow.com/a/75552881/764870), it must be enabled in the registry to work (Set `HKLM\SYSTEM\CurrentControlSet\Services\Tcpip\Parameters\DisableUserTOSSetting` to 0 and restart the machine). Should we document it somewhere?

File: src/tool_getparam.c
Comment (orgads) [reply src/tool_getparam.c]: Replaced with `IP_TOS`. Looks like (at least on FreeBSD) `netinet/in.h` has to be included before `netinet/ip.h` anyway.

Let's see if it compiles on all platforms.

Comment (orgads) [reply src/tool_getparam.c]: Ok, no more macros. I just used the numeric values inline (they are invariant).

**Comment by orgads** [comment]
The failing test in `quiche` looks like a flaky test, I doubt it's related to my changes.

File: docs/cmdline-opts/type-of-service.md
Comment (bagder) [inline docs/cmdline-opts/type-of-service.md:21]:
```suggestion
The values allowed for \<string\> can be a numeric value between 1 and 255 or one of the
```

File: docs/cmdline-opts/type-of-service.md
Comment (bagder) [inline docs/cmdline-opts/type-of-service.md:52]:
```suggestion
* MINCOST

This option is limited to IPv4 connections only. When curl uses IPv6, this option does nothing.
```

File: docs/options-in-versions
Comment (bagder) [inline docs/options-in-versions:260]: It is too late for 8.8.0 but it should be fit for 8.9.0.
```suggestion
--type-of-service                    8.9.0
```

File: docs/cmdline-opts/type-of-service.md
Comment (bagder) [inline docs/cmdline-opts/type-of-service.md:19]:
```suggestion
Set IP Type of Service (TOS). (Added in 8.9.0).
```

File: src/tool_operate.c
Comment (orgads) [reply src/tool_operate.c:173]: Can you suggest a way to check the protocol type here?

File: docs/cmdline-opts/type-of-service.md
Comment (orgads) [reply docs/cmdline-opts/type-of-service.md:21]: Done

File: docs/cmdline-opts/type-of-service.md
Comment (orgads) [reply docs/cmdline-opts/type-of-service.md:52]: Done

File: docs/options-in-versions
Comment (orgads) [reply docs/options-in-versions:260]: Done

File: docs/cmdline-opts/type-of-service.md
Comment (orgads) [reply docs/cmdline-opts/type-of-service.md:19]: Done

File: src/tool_operate.c
Comment (orgads) [reply src/tool_operate.c:2235]: Ok

File: src/tool_operate.c
Comment (bagder) [reply src/tool_operate.c]: Can `getsockname` help?

Comment (orgads) [reply src/tool_operate.c]: Actually I found `getsockopt(sockfd, SOL_SOCKET, SO_DOMAIN, &family, &len)`.

But before I change that, isn't it better to show an error on such failure rather than silently ignoring the option?

Comment (bagder) [reply src/tool_operate.c]: This goes back to what I asked you from the start: users typically won't know if IPv4 or IPv6 gets used so they can't tell when this option takes effect. Do you think a user wants an error message when it doesn't? I doubt that.

Since v4/v6 may even vary between separate invokes to the same host name it could even result in an error message intermittently every nth invoke which probably will be highly confusing and hardly productive for users.

Comment (orgads) [reply src/tool_operate.c]: Done

File: src/tool_getparam.c
Comment (rpigott) [inline src/tool_getparam.c:1037]: Not that it really matters, but TOS/DSCP constants are also available from `#include <netinet/ip.h>` if you want to use them.

There you should also find the constants IPTOS_DSCP_VA (0x0b, RFC5865 Voice Admit) and IPTOS_DSCP_LE (0x04, RFC8622 Lower Effort) defined, which I don't see represented in this table. The full list is https://www.iana.org/assignments/dscp-registry/dscp-registry.xhtml#dscp-registry-1.

File: src/tool_getparam.c
Comment (rpigott) [inline src/tool_getparam.c:1058]: Are you sure it even makes sense to allow ECT(0) and ECT(1) values?  Linux masks these bits out for SOCK_STREAM sockets anyway.

https://github.com/torvalds/linux/blob/29c73fc794c83505066ee6db893b2a83ac5fac63/net/ipv4/ip_sockglue.c#L590-L592

File: docs/cmdline-opts/type-of-service.md
Comment (rpigott) [inline docs/cmdline-opts/type-of-service.md:47]: This value isn't present in the table in the code. I'm not sure it's worth supporting anyway, though.

**Review by rpigott — COMMENTED** [review]
I'm also interested in this feature, so thanks for writing it up!

File: src/tool_getparam.c
Comment (orgads) [reply src/tool_getparam.c:1037]: Thank you. I initially used the implementation from #7762, but then the tests failed on some platforms that don't have `netinet/ip.h`, so I ended up just hardcoding them.

Thanks for the list, I'll update it.

File: docs/cmdline-opts/type-of-service.md
Comment (orgads) [reply docs/cmdline-opts/type-of-service.md:48]: Added now. Thank you.

**Comment by orgads** [comment]
@bagder Is this good to go, now that 8.8 was released?

**Review by bagder — COMMENTED** [review]
We cannot merge this anyway until the feature window opens on June 1st if everything goes well.

File: src/tool_operate.c
Comment (orgads) [reply src/tool_operate.c:2246]: Done

File: src/tool_operate.c
Comment (orgads) [reply src/tool_operate.c:180]: Done

File: src/tool_operate.c
Comment (vszakats) [inline src/tool_operate.c:2262]: Would it make sense to display a warning or trace message if a >0 `ip_tos` was requested by the user, but the feature isn't supported?

Comment (orgads) [reply src/tool_operate.c]: Done

**Comment by orgads** [comment]
Is the msys2 with debug flaky? I couldn't find anything in the log that looks related to my changes.

**Comment by vszakats** [comment]
@orgads: Yes, the MSYS2 mingw-w64 ones sometimes **hang** in tests. I'm tracking the problem here: https://github.com/curl/curl/pull/13599#issuecomment-2119372376. Could not yet figure out _which_ test is prone to hanging though.

File: docs/cmdline-opts/type-of-service.md
Comment (bagder) [inline docs/cmdline-opts/type-of-service.md:19]: Since you added  support for IPv6 and its "Traffic Class" field:

1. Should this option really be named like this now?
2. Should the documentation not specify the IPv6 details?

File: docs/cmdline-opts/type-of-service.md
Comment (bagder) [inline docs/cmdline-opts/type-of-service.md:56]: This seems to be wrong now?

File: docs/cmdline-opts/type-of-service.md
Comment (bagder) [inline docs/cmdline-opts/type-of-service.md:54]: Are all these values also specified/used for IPv6?

File: docs/cmdline-opts/type-of-service.md
Comment (orgads) [reply docs/cmdline-opts/type-of-service.md:19]: 1. Do you have a better suggestion? Maybe add another `--traffic-class` which does the same?
2. These fields have the same purpose and structure (6 bits DSCP, 2 bits ECN), only the name and the location within the packet differs.

File: docs/cmdline-opts/type-of-service.md
Comment (orgads) [reply docs/cmdline-opts/type-of-service.md:56]: Thanks! Removed the note.

File: docs/cmdline-opts/type-of-service.md
Comment (orgads) [reply docs/cmdline-opts/type-of-service.md:54]: Yes, it's the same.

**Comment by orgads** [comment]
Is there a way to retrigger only the failing job?

**Comment by vszakats** [comment]
@orgads: I don't think that is made possible by GitHub. Tried to fix the obscure issue that prevented caching the failing download here: 1d63e331de4ef8ad1e2f228c6815fa6fecec6dbb.

Restarted that job manually — if it fails again, please rebase on master.

**Comment by orgads** [comment]
Passed, thanks!

File: docs/cmdline-opts/type-of-service.md
Comment (bagder) [reply docs/cmdline-opts/type-of-service.md]: No, this should be a single option I think. Having two different ones doing the same thing does not make anything easier for anyone. It would make more sense to come up with a generic name that works for applying to both cases. Maybe it should even include "IP" in the name to better hint what it is about.

Maybe...

- `--ip-traffic`
- `--ip-type`
- `--ip-service`

?

Comment (orgads) [reply docs/cmdline-opts/type-of-service.md]: I'm not sure any of these suggestions is more descriptive than type-of-service.

TOS used to have a different meaning, but nowadays, it is just like traffic class, divided to 6 bits DSCP and 2 bits ECN.

Maybe we should just have `--ip-dscp` and `--ip-ecn`?

Comment (rpigott) [reply docs/cmdline-opts/type-of-service.md]: Just some helpful info on other tools:

linux iproute2 has `tos` and its alias `dsfield` for adding/showing routes that match on the dsfield. It accepts a named DSCP value or a masked 8 bit TOS value, and the ECN bits are ignored. The option is named the same for both v4 and v6 routing tables. E.g. `ip route add to 1.2.3.4/32 dsfield AF21 counter`

linux nftables has `ip dscp` and `ip6 dscp` payload expresions that match/set the 6 bit ds field in the respective protocol family. It accepts either a named or 6 bit DSCP value. E.g. `nft add rule inet filter input ip dscp af21 meta priority set 1:4`.

`tcpdump -v` displays the tos / tclass byte as "tos" or "class" followed by the 8 bit hex value of the byte, including the ECN field.

`ping -Q` sets the tos / tclass byte just like this patch, and linux copies the entire byte in echo replies.

`--ip-tos` or even just `--tos` seems fine to me. I'm partial to `--dscp` I think. I mentioned it earlier but  linux ignores the ECN field from this sockopt for SOCK_STREAM sockets, and I suppose the quic library will manage the ECN field otherwise, so I expect that setting the the ECN field with this option will never be effective, hence `--dscp` seems reasonable.

Comment (orgads) [reply docs/cmdline-opts/type-of-service.md]: @rpigott Thanks a lot for the detailed overview!

I'm a bit reluctant. My considerations:
* `tos`/`type-of-service`:
  * Pro: Widely used (relatively for this niche feature :)) and should be easily recognized.
  * Con: It's an obsolete name for this field, and doesn't describe its actual use.
* `traffic-class`:
  * Pro: Better description of the field, was chosen over ToS for IPv6, which is more recent.
  * Con: Less known, harder to discover.

I call for a vote.

❤ `--ip-tos`
🚀 `--tos`
🎉 `--ip-type-of-service`
👀 `--ip-traffic-class`

Comment (orgads) [reply docs/cmdline-opts/type-of-service.md]: Aaaaaaand we have a winner!

Changed to `--ip-tos`. Ready for review.

File: docs/cmdline-opts/ip-tos.md
Comment (bagder) [inline docs/cmdline-opts/ip-tos.md:6]:
```suggestion
Help: Set IP Type of Service or Traffic Class
```

We want this phrase to be as short and succinct as possible

Comment (orgads) [reply docs/cmdline-opts/ip-tos.md]: Done

**Review by bagder — APPROVED** [review]

**Comment by orgads** [comment]
All passed. I have one last PR, but it depends on this one, so I'll push it after this is merged.

Thanks you all for your constructive review!

**Review by vszakats — APPROVED** [review]

File: src/tool_getparam.c
Comment (vszakats) [inline src/tool_getparam.c:1684]: Just curious, is declaring a variable in a `case` block valid C89? I was under the impression that it isn't, and curl also used explicit blocks in some other code in similar cases.

Comment (orgads) [reply src/tool_getparam.c:1684]: It's in a new block.

Comment (orgads) [reply src/tool_getparam.c:1684]: Declaring *and initializing* a variable in a case that has subsequent cases is considered undefined behavior (in all C/C++ variants), and GCC warns about it.

Comment (vszakats) [reply src/tool_getparam.c:1684]: It'd make sense, but the pattern often (?) seen is this:
```c
case <n>:
{
   int var = 1;
}
```
`case` being a block is not 100% self-evident: https://en.wikipedia.org/wiki/Duff%27s_device

Hence the question.

Comment (orgads) [reply src/tool_getparam.c:1684]: But there's a brace there. The same style is used all over this file. Examples:
```c
case C_PARALLEL_MAX: {  /* --parallel-max */
case C_LOCAL_PORT: { /* --local-port */
case C_REFERER: { /* --referer */
```

Comment (vszakats) [reply src/tool_getparam.c:1684]: I apologize for the noise, indeed there is, I didn't notice!

**Comment by orgads** [comment]
Is anything still missing, or are we good to go?

**Comment by bagder** [comment]
Thanks!

**Pull request closed without merging**
